src/hyperv/storvsc/storvsc.c

```diff
@@ -1050,14 +1048,27 @@ static void create_storvsc_request(struct storvsc_hcb *hcb, struct hv_storvsc_re
 
     reqp->hcb = hcb;
 
-    if (hcb->data_len) {
-        struct storvsc_gpa_range *prplist = &reqp->prp_list;
+    uint32_t bytes_to_copy = hcb->data_len;
+    struct storvsc_gpa_range *prplist = &reqp->prp_list;
+    if (bytes_to_copy != 0) {
+        u64 phys_addr = physical_from_virtual(&hcb->data_addr[hcb->data_len - bytes_to_copy]);
+        assert(phys_addr != INVALID_PHYSICAL);
         prplist->gpa_range.gpa_len = hcb->data_len;
-        prplist->gpa_range.gpa_ofs = hcb->data_addr & PAGEMASK;
-        assert(prplist->gpa_range.gpa_ofs == 0);
-        prplist->gpa_page[0] = hcb->data_addr >> PAGELOG;
-        reqp->prp_cnt = 1;
+        prplist->gpa_range.gpa_ofs = phys_addr & PAGEMASK;
     }
+
+    uint32_t pfn_num = 0;
+    while (bytes_to_copy != 0) {
+        u64 phys_addr = physical_from_virtual(&hcb->data_addr[hcb->data_len - bytes_to_copy]);
```

Comment:
This physical_from_virtual was already done above. Maybe change to do - while and move under "if (bytes_to_copy != 0)" block.